### ISSUE TITLE
s3 (system/3) deposit aaaa A0-A9 is converted to 36h

### ISSUE THREAD

**Issue by mstram** [simh/simh#527]
With the S3 simulator, a d(eposit) of any value from A0-A9 is being converted to 36h.

AA-AF is deposited without conversion.
Even with using -h,  A0-A3 is still  converted to 36h.       ( d -h 0A0) prefix 0 is still necessary.
Prefixing with 0 makes it work   (d xxxx 0A0 ....   d xxxx 0A9).
Or using -o (octal)   d  -o xxxx 240       d -o xxxx 251   also works.
dep Bx, Cx, Dx, Ex, Fx   all work

Maybe something in that deposits for the S3 default to hex values ?

Code getting confused by leading 'A' ? (alpha ?

Mike

**Comment by markpizz** [comment]
Stepping through the code which parses the 'failing' command suggests that the leading A followed by a digit is being matched with a machine opcode "A".  I'm not familiar with this machine at all, but that might be a legitimate textual instruction format for the A instruction.  Other simulators explicitly specify/require a -m switch when deposit data is to be parsed as instructions.  It seems that this simulator is trying to determine that its input is on the fly.  If this parsing is broken, then the problem lies in the s3_sys.c/parse_sym routine.  Feel free to suggest a fix that does the right thing for you and doesn't break anything else.

**Comment by mstram** [comment]
Hi Mark,

Yes 'A' is a valid mnemonic ( Add to Register) page 2-3 in ref manual
linked below.

According to the manual  there are no legal mnemonics that include a
non-alpha char.
( Are there *any* assemblers with a numeric any of their mnemonics ?)

So I removed the || isdigit(*cptr) from    s3_sys.c/parse_sym  ~line 528

    if (*cptr == ',' || *cptr == '\0' || *cptr == '('  )

It seems to be working.

Here's the reference manual :

http://bitsavers.org/pdf/ibm/system3/GA21-9236-1_IBM_System-3_Model_8_10_12_15_Components_Reference_Manual_Nov77.pdf

Mike

**Comment by markpizz** [comment]
True, but, in the end, you will still have to enter 0A instead of merely A to represent that value.  The real problem seems to be the overloaded parsing model.  

A more regular solution would seem to be to follow the model used in other simulators and only parse for opcodes when explicitly specified with the -M switch.

**Comment by markpizz** [comment]
Like other simulators, this simulator already uses the -M switch when a user wants to display the contents of memory a instructions.  To me, it sounds more reasonable to require the -M switch.  I'm thinking that is more reasonable AS LONG AS there aren't a ton of examples in documentation or existing scripts somewhere that explicitly deposit without a -M switch.  If such examples exist, then they should either be changed, or the parser needs to be made more robust so that it silently accepts ABSOLUTELY unambiguous input and emits warnings when ambiguities exist (the case you started with is certainly one, but there more than likely are others).

**Comment by mstram** [comment]
> Other simulators explicitly specify/require a -m switch when deposit data is to be parsed as instructions. I

Well, FWIW, I just tested all the sims.

All except the IBM1130, and the swtp6800mp-a.exe , swtp6800mp-a2.exe accept alpha instructions WITHOUT the -m switch ;)

**Comment by markpizz** [comment]
OK.  I only checked about a dozen simulators and, indeed there isn't a -m switch used when parsing input, so that advice was in the wrong direction.  Meanwhile, what I did notice was rigorous parsing of the details of instruction structure with error exits (SCPE_ARG) when unreasonable instruction input was seen.  Sometimes these exits returned immediately, other times alternative parsing (non-instruction) was then attempted.  These observations were more along the line of the back half of my suggestion:

> ... or the parser needs to be made more robust so that it silently accepts ABSOLUTELY unambiguous input and emits warnings when ambiguities exist (the case you started with is certainly one, but there more than likely are others).

I'm thinking that more rigor than merely removing the isdigit() check in the conditional that you tried earlier would really be necessary.
